dc/s2n-quic-dc/src/stream/environment/tokio/pool.rs

```diff
@@ -0,0 +1,234 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use super::udp::{ApplicationSocket, RecvSocket, WorkerSocket};
+use crate::{
+    event,
+    path::secret::Map,
+    socket::recv::{pool::Pool as Packets, router::Router, udp},
+    stream::{
+        environment::tokio::Environment,
+        recv::dispatch::{Allocator as Queues, Control, Stream},
+        socket::{application::Single, fd::udp::CachedAddr, SendOnly, Tracing},
+    },
+    sync::ring_deque::Capacity,
+};
+use s2n_quic_platform::socket::options::{Options, ReusePort};
+use std::{
+    io,
+    io::Result,
+    net::UdpSocket,
+    sync::{
+        atomic::{AtomicUsize, Ordering},
+        Arc, Mutex,
+    },
+};
+use tokio::io::unix::AsyncFd;
+use tracing::Instrument;
+
+#[derive(Clone)]
+#[non_exhaustive]
+pub struct Config {
+    pub blocking: bool,
+    pub reuse_port: bool,
+    pub stream_queue: Capacity,
+    pub control_queue: Capacity,
+    pub max_packet_size: u16,
+    pub packet_count: usize,
+    pub workers: Option<usize>,
+    pub map: Map,
+}
+
+impl Config {
+    pub fn new(map: Map) -> Self {
+        Self {
+            blocking: false,
+            reuse_port: false,
+            // TODO tune these defaults
+            stream_queue: Capacity {
+                max: 4096,
+                initial: 256,
+            },
+
+            // set the control queue depth shallow, since we really only need the most recent ones
+            control_queue: Capacity { max: 8, initial: 8 },
+
+            // Allocate 1MB at a time
```

Comment:
Hm, we might regret this a bit -- for the "many peers" testing we're doing, it's important that the per-peer usage is quite small. E.g., if you want ~20k peers on a single host then 1MB/peer is very expensive. Probably not blocking this, but I think we should think about how we can mitigate that... maybe support a "low memory" mode that's less aggressive on the buffer allocation or something?

Reply:
Note that this is wouldn't be per peer - it's per socket worker. So it should only be maxing out at how many packets are received and being processed at a time.

But, yes, we will need to do some tuning to these numbers. The problem we GRO is you have to provide a full 2^16 payload, otherwise packets get truncated. And that gets tricky to manage with the packet allocator since if you just get a single 50 byte packet, then the rest of the space for that descriptor needs to wait for the packet to be freed.

Reply:
But I wonder if we do some copy optimizations where if the payloads are under a certain size it might be cheaper to copy them into a smaller space rather than holding up the entire 2^16 bytes? not sure.

Reply:
I mean per dc endpoint (basically having 1000s of path secret maps in a single process - not really useful outside testing, but testing is still important).